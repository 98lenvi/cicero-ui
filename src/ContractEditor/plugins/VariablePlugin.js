import React from 'react';
import PropTypes from 'prop-types';

/**
 * A plugin for a variable
 */
function VariablePlugin() {
  const plugin = 'Variable';
  const tags = ['variable'];
  const markdownTags = ['variable'];
  const schema = {
    inlines: {
      variable: {
        nodes: [
          {
            match: [{ type: 'text' }],
          },
        ],
      },
    },
  };

  /**
     * @param {Event} event
     * @param {Editor} editor
     * @param {Function} next
     */
  function onEnter(event, editor, next) {
    return next();
  }

  /**
 * Handles change to document.
 */
  function onChange(editor, next) {
    console.log('onChange - VariablePlugin');
    let variableNode = null;
    const variables = editor.value.inlines.some(inline => inline.type === 'variable');
    console.log('variables', variables);
    if (variables.size === 1) {
      variableNode = variables.get(0);
    }

    if (!variableNode) {
      return next();
    }

    console.log(variableNode);
    const { name } = variableNode.data;

    if (name) {
      console.log(`Found variable on node: ${name}`);
    } else {
      console.log('Variable name not found on node.');
    }


    return next();
  }

  /**
     * @param {Event} event
     * @param {Editor} editor
     * @param {Function} next
     */
  function onKeyDown(event, editor, next) {
    switch (event.key) {
      case 'Enter':
        return onEnter(event, editor, next);
      default:
        return next();
    }
  }

  /**
   * Render a Slate inline.
   *
   * @param {Object} props
   * @param {Editor} editor
   * @param {Function} next
   * @return {Element}
   */
  function renderInline(props, editor, next) {
    const { attributes, children, node } = props;

    switch (node.type) {
      case 'variable': {
        const { data } = node;
        const name = data.get('name');
        return (
          <a {...attributes} href={name}>
            {children}
          </a>
        );
      }

      default: {
        return next();
      }
    }
  }

  /**
     * @param {ToMarkdown} parent
     * @param {Node} value
     */
  function toMarkdown(parent, value) {
    let markdown = `<variable ${value.data.get('attributeString')}>`;

    value.nodes.forEach((li) => {
      const text = parent.recursive(li.nodes);
      markdown += text;
    });

    markdown += '</variable>\n\n';
    return markdown;
  }

  /**
 * Handles data from markdown.
 */
  function fromMarkdown(stack, event, tag) {
    const block = {
      object: 'inline',
      type: 'variable',
      data: Object.assign(tag),
      nodes: [],
    };

    stack.push(block);

    stack.addTextLeaf({
      object: 'leaf',
      text: tag.content ? tag.content : '',
      marks: [],
    });
    stack.pop();
    console.log(stack);
    return true;
  }

  /**
 * Handles data from the HTML format.
 */
  function fromHTML(editor, el, next) {
    return {
      object: 'inline',
      type: 'variable',
      data: {},
      nodes: next(el.childNodes),
    };
  }

  return {
    plugin,
    tags,
    markdownTags,
    schema,
    onKeyDown,
    renderInline,
    toMarkdown,
    fromMarkdown,
    fromHTML,
    onChange,
  };
}

export default VariablePlugin;
