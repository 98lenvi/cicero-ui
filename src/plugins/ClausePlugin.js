import React from 'react';
import { Point } from 'slate';
import PropTypes from 'prop-types';
import { Icon } from 'semantic-ui-react';
import styled from 'styled-components';
import { Template, Clause } from '@accordproject/cicero-core';
import '../styles.css';
// eslint-disable-next-line camelcase
import murmurhash3_32_gc from './murmurhash3_gc';

const StyledIcon = styled(Icon)`
  color: #ffffff !important;
`;

/**
 * A plugin for a clause embedded in a contract
 * @param {*} ClauseComponent - a react component used to render the clause
 * @param {*} customLoadTemplate - a custom function used to load templates
 * @param {*} customParseClause - a custom function used to parse clause text
 */
function ClausePlugin(ClauseComponent, customLoadTemplate, customParseClause) {
  const plugin = 'Clause';
  const tags = ['clause'];
  const markdownTags = ['clause'];

  const templates = {};
  const schema = {
    blocks: {
      Clause: {
        nodes: [
          {
            match: [{ type: 'paragraph' }],
          },
        ],
      },
    },
  };

  /**
   * Called by the clause plugin into the contract editor
   * when we need to load a template
   */
  async function loadTemplate(templateUri) {
    let template = templates[templateUri];
    if (!template) {
      console.log(`Loading template: ${templateUri}`);
      template = await Template.fromUrl(templateUri);
      templates[templateUri] = template;
    }
  }

  /**
   * Called by the clause plugin into the contract editor
   * when we need to parse a clause
   */
  function parseClause(templateUri, text, clauseId) {
    try {
      const template = templates[templateUri];
      if (template) {
        console.log(`parseClause: ${templateUri} with ${text}`);
        const clause = new Clause(template);
        clause.parse(text);
        return Promise.resolve(clause.getData());
      }
      return Promise.resolve('Template not loaded.');
    } catch (err) {
      return Promise.resolve(err);
    }
  }

  const loadTemplateCallback = customLoadTemplate || loadTemplate;
  const parseClauseCallback = customParseClause || parseClause;

  /**
     * @param {Event} event
     * @param {Editor} editor
     * @param {Function} next
     */
  function onEnter(event, editor, next) {
    return next();
  }

  /**
   * Adds an annotation to the editor
   *
   * @param {*} editor
   * @param {*} annotation
   */
  function addAnnotation(editor, annotation) {
    editor.addAnnotation(annotation);
  }

  /**
   * Remove the parse annotations from the editor
   *
   * @param {*} editor
   */
  function removeParseAnnotations(editor) {
    const { annotations } = editor.value;

    editor.withoutSaving(() => {
      annotations.forEach((ann) => {
        if (ann.type.startsWith('parse')) {
          editor.removeAnnotation(ann);
        }
      });
    });
  }

  /**
   * Checks whether an annotation exists
   *
   * @param {*} editor
   * @param {*} annotation
   * @returns {boolean} true if the annotation already exists on the value
   */
  function annotationExists(editor, annotation) {
    const { annotations } = editor.value;
    return annotations.get(annotation.key);
  }

  /**
  * Handles change to document.
  */
  function onChange(editor, next) {
    console.log('onChange');
    let clauseNode = null;
    let textNode = null;
    const textNodes = editor.value.texts;
    if (textNodes.size === 1) {
      textNode = textNodes.get(0);
      const para = editor.value.document.getParent(textNode.key);
      clauseNode = editor.value.document.getParent(para.key);
    }

    if (!clauseNode || clauseNode.type !== 'clause') {
      return next();
    }

    const nodeAttributes = clauseNode.data.get('attributes');
    const { src, clauseid } = nodeAttributes;
    const { selection } = editor.value;

    const anchor = Point.create(selection.anchor).moveToStartOfNode(clauseNode)
      .normalize(editor.value.document);
    const focus = Point.create(selection.focus).moveToEndOfNode(clauseNode)
      .normalize(editor.value.document);
    const parseText = textNode.text.trim();
    // @ts-ignore
    const hash = murmurhash3_32_gc(`${parseText} at ${anchor.toJSON()} to ${focus.toJSON()}`, 0xdeadbeef);

    const annotation = {
      anchor,
      focus,
      key: `clauseParse-${hash}`
    };

    // we only re-parse and modify the value if the text has changed
    if (!annotationExists(editor, annotation)) {
      removeParseAnnotations(editor);
      parseClauseCallback(src, parseText, clauseid)
        .then((parseResult) => {
          annotation.type = 'parseResult';
          annotation.data = parseResult;
          addAnnotation(editor, annotation);
        })
        .catch((error) => {
          annotation.type = 'parseError';
          annotation.data = { error };
          addAnnotation(editor, annotation);
        });
    }

    return next();
  }

  /**
  * @param {Event} event
  * @param {Editor} editor
  * @param {Function} next
  */
  function onKeyDown(event, editor, next) {
    switch (event.key) {
      case 'Enter':
        return onEnter(event, editor, next);
      default:
        return next();
    }
  }

  /**
  * @param {Object} props
  * @param {Editor} editor
  * @param {Function} next
  */
  function renderBlock(props, editor, next) {
    renderBlock.propTypes = {
      node: PropTypes.any,
      attributes: PropTypes.any,
      children: PropTypes.any,
    };

    const { node, children } = props;

    const nodeAttributes = node.data.get('attributes');
    const { src, clauseId } = nodeAttributes;

    if (src) {
      console.log(`handing over responsibility of loading: ${src}`);
      loadTemplateCallback(src.toString());
    }

    return (<ClauseComponent templateUri={src} clauseId={clauseId} {...props}>{children}</ClauseComponent>);
  }

  /**
   * @param {ToMarkdown} parent
   * @param {Node} value
   */
  function toMarkdown(parent, value) {
    let markdown = `\n\n<clause src=${value.data.get('attributes').src} id=${value.data.get('attributes').id}>`;

    value.nodes.forEach((li) => {
      const text = parent.recursive(li.nodes);
      markdown += text;
    });

    markdown += '</clause>';
    return markdown;
  }

  /**
  * Handles data from markdown.
  */
  function fromMarkdown(stack, event, tag) {
    const block = {
      object: 'block',
      type: 'clause',
      data: Object.assign(tag),
      nodes: [],
    };

    stack.push(block);

    const para = {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [],
    };

    stack.push(para);
    stack.addTextLeaf({
      object: 'leaf',
      text: tag.content ? tag.content : '',
      marks: [],
    });
    stack.pop();
    stack.pop();
    return true;
  }

  /**
  * Handles data from the HTML format.
  */
  function fromHTML(editor, el, next) {
    return {
      object: 'block',
      type: 'clause',
      data: {},
      nodes: next(el.childNodes),
    };
  }

  /**
  * Handles a button click.
  */
  function onClickButton(editor, event) {
    event.preventDefault();
    alert('Clause plugin button clicked!');
  }

  /**
   * Render a clause toolbar button.
   *
   * @param {Editor} editor
   * @return {Element}
   */
  function renderToolbar(editor) {
    return (<StyledIcon
      key={plugin}
      name='legal'
      aria-label='clause'
      onMouseDown={event => onClickButton(editor, event)}
    />);
  }

  /**
   * Render Slate annotations.
   */
  function renderAnnotation(props, editor, next) {
    const { children, annotation, attributes } = props;

    switch (annotation.type) {
      case 'parseError':
        return (
          <span {...attributes} className='parseError'>
            {children}
          </span>
        );
      default:
        return next();
    }
  }

  return {
    plugin,
    tags,
    markdownTags,
    schema,
    onKeyDown,
    renderBlock,
    toMarkdown,
    fromMarkdown,
    fromHTML,
    onChange,
    renderToolbar,
    renderAnnotation
  };
}

export default ClausePlugin;
