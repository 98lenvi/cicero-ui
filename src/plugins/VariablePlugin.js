import React from 'react';

/**
 * A plugin for a variable
 */
function VariablePlugin() {
  const name = 'variable';

  const tags = [
    {
      html: 'variable',
      slate: 'variable',
      md: 'variable'
    }
  ];

  /**
   * Augment the base schema with the variable type
   * @param {*} schema
   */
  const augmentSchema = ((schema) => {
    const additions = {
      inlines: {
        variable: {
          nodes: [{
            match: { object: 'text' }
          }]
        },
      },
    };

    const newSchema = JSON.parse(JSON.stringify(schema));
    newSchema.inlines = { ...newSchema.inlines, ...additions.inlines };
    newSchema.document.nodes[0].match.push({ type: tags[0].slate });
    newSchema.blocks.paragraph.nodes[0].match.push({ type: tags[0].slate });
    return newSchema;
  });

  /**
   * Only allows words that don't have marks to be edited
   *
   * @param {Value} value - the Slate value
   */
  const isEditable = ((value, code) => {
    const inVariable = value.inlines.size > 0 && value.inlines.every(node => node.type === 'variable');
    const { anchor } = value.selection;
    console.log(anchor.toJSON());

    if (code === 'backspace') {
      if (inVariable) {
        // if we hit backspace and are at the zeroth
        // position of a variable prevent deleting the char
        // that precedes the variable
        return anchor.offset > 0;
      }

      // if we hit backspace and are outside of a variable
      // allow deleting the last char of the variable
      // IFF the variable has more than 1 char
      const prev = value.document.getPreviousSibling(anchor.path);
      return prev && anchor.offset === 0 && prev.type === 'variable' && prev.getFirstText().text.length > 1;
    }

    if (code === 'input') {
      // if are outside of a variable allowing
      // extending the variable
      const prev = value.document.getPreviousSibling(anchor.path);
      return prev && anchor.offset === 0 && prev.type === 'variable';
    }

    return inVariable;
  });

  /**
   * Render a Slate inline.
   *
   * @param {Object} props
   * @param {Editor} editor
   * @param {Function} next
   * @return {Element}
   */
  function renderInline(props, editor, next) {
    const { attributes, children, node } = props;

    switch (node.type) {
      case 'variable': {
        const { data } = node;
        const value = data.get('value');
        // @ts-ignore
        return <span {...attributes} className='variable'>
            {children}
          </span>;
      }

      default: {
        return next();
      }
    }
  }

  /**
     * @param {ToMarkdown} parent
     * @param {Node} value
     */
  function toMarkdown(parent, value) {
    return `<variable ${value.data.get('attributeString')}/>\n\n`;
  }

  /**
 * Handles data from markdown.
 */
  function fromMarkdown(stack, event, tag, node) {
    const parent = stack.peek();

    // variables can only occur inside paragraphs
    if (!parent.type || parent.type !== 'paragraph') {
      const para = {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [],
      };
      stack.push(para);
    }

    const inline = {
      object: 'inline',
      type: 'variable',
      data: Object.assign(tag),
      nodes: [{
        object: 'text',
        text: `${tag.attributes.value}`,
      }]
    };

    stack.append(inline);

    if (!parent.type || parent.type !== 'paragraph') {
      stack.pop();
    }

    return true;
  }

  /**
 * Handles data from the HTML format.
 */
  function fromHTML(editor, el, next) {
    return {
      object: 'block',
      type: 'variable',
      data: {},
      nodes: next(el.childNodes),
    };
  }

  return {
    name,
    tags,
    augmentSchema,
    isEditable,
    renderInline,
    toMarkdown,
    fromMarkdown,
    fromHTML,
  };
}

export default VariablePlugin;
