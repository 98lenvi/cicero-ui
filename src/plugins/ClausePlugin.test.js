import { PluginManager, List, ToMarkdown } from '@accordproject/markdown-editor';
import { Value } from 'slate';

import ClausePlugin from './ClausePlugin';
import VariablePlugin from './VariablePlugin';

let toMarkdown = null;

beforeAll(() => {
  const clausePlugin = ClausePlugin(null, null);
  const pluginManager = new PluginManager([List(), clausePlugin, VariablePlugin()]);
  toMarkdown = new ToMarkdown(pluginManager);
});

const slateValueJSON = {
  object: 'value',
  document: {
    object: 'document',
    nodes: [
      {
        object: 'block',
        type: 'clause',
        data: {
          tag: 'clause',
          attributes: {
            src: 'ap://acceptance-of-delivery@0.12.0#b8eae2fb3c2571284b4dcfc1fc348d8234d0c5db158962043720828be2eb9085',
            clauseid: 'd3cdb02e-6982-47a1-9ba3-350d46212b63'
          },
          attributeString: 'src = "ap://acceptance-of-delivery@0.12.0#b8eae2fb3c2571284b4dcfc1fc348d8234d0c5db158962043720828be2eb9085"clauseid = "d3cdb02e-6982-47a1-9ba3-350d46212b63"',
          content: '',
          closed: false
        },
        nodes: [
          {
            object: 'block',
            type: 'paragraph',
            data: {

            },
            nodes: [
              {
                object: 'block',
                type: 'paragraph',
                data: {

                },
                nodes: [
                  {
                    object: 'text',
                    text: 'Acceptance of Delivery. ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'shipper',
                        value: '%22Party%20A%22'
                      },
                      attributeString: 'id = "shipper"value = "%22Party%20A%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party A"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' will be deemed to have completed its delivery obligations if in ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'receiver',
                        value: '%22Party%20B%22'
                      },
                      attributeString: 'id = "receiver"value = "%22Party%20B%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party B"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: "'s opinion, the ",
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'deliverable',
                        value: '%22Widgets%22'
                      },
                      attributeString: 'id = "deliverable"value = "%22Widgets%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Widgets"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' satisfies the Acceptance Criteria, and ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'receiver',
                        value: '%22Party%20B%22'
                      },
                      attributeString: 'id = "receiver"value = "%22Party%20B%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party B"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' notifies ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'shipper',
                        value: '%22Party%20A%22'
                      },
                      attributeString: 'id = "shipper"value = "%22Party%20A%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party A"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' in writing that it is accepting the ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'deliverable',
                        value: '%22Widgets%22'
                      },
                      attributeString: 'id = "deliverable"value = "%22Widgets%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Widgets"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: '.',
                    marks: [

                    ]
                  }
                ]
              },
              {
                object: 'block',
                type: 'paragraph',
                data: {

                },
                nodes: [
                  {
                    object: 'text',
                    text: 'Inspection and Notice. ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'receiver',
                        value: '%22Party%20B%22'
                      },
                      attributeString: 'id = "receiver"value = "%22Party%20B%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party B"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' will have ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'businessDays',
                        value: '10'
                      },
                      attributeString: 'id = "businessDays"value = "10"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '10',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: " Business Days' to inspect and evaluate the ",
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'deliverable',
                        value: '%22Widgets%22'
                      },
                      attributeString: 'id = "deliverable"value = "%22Widgets%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Widgets"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' on the delivery date before notifying ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'shipper',
                        value: '%22Party%20A%22'
                      },
                      attributeString: 'id = "shipper"value = "%22Party%20A%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party A"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' that it is either accepting or rejecting the ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'deliverable',
                        value: '%22Widgets%22'
                      },
                      attributeString: 'id = "deliverable"value = "%22Widgets%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Widgets"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: '.',
                    marks: [

                    ]
                  }
                ]
              },
              {
                object: 'block',
                type: 'paragraph',
                data: {

                },
                nodes: [
                  {
                    object: 'text',
                    text: 'Acceptance Criteria. The "Acceptance Criteria" are the specifications the ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'deliverable',
                        value: '%22Widgets%22'
                      },
                      attributeString: 'id = "deliverable"value = "%22Widgets%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Widgets"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' must meet for the ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'shipper',
                        value: '%22Party%20A%22'
                      },
                      attributeString: 'id = "shipper"value = "%22Party%20A%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Party A"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ' to comply with its requirements and obligations under this agreement, detailed in ',
                    marks: [

                    ]
                  },
                  {
                    object: 'inline',
                    type: 'variable',
                    data: {
                      tag: 'variable',
                      attributes: {
                        id: 'attachment',
                        value: '%22Attachment%20X%22'
                      },
                      attributeString: 'id = "attachment"value = "%22Attachment%20X%22"',
                      content: '',
                      closed: true
                    },
                    nodes: [
                      {
                        object: 'text',
                        text: '"Attachment X"',
                        marks: [

                        ]
                      }
                    ]
                  },
                  {
                    object: 'text',
                    text: ', attached to this agreement.',
                    marks: [

                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
};

test('can convert slate with variables', () => {
  const slateValue = Value.fromJSON(slateValueJSON);
  const value = toMarkdown.recursive(slateValue.document.nodes);
  expect(value).toMatchSnapshot();
});
