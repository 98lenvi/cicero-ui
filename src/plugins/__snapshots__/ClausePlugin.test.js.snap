// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can convert slate with variables 1`] = `
"

\`\`\` <clause src=ap://acceptance-of-delivery@0.12.0#b8eae2fb3c2571284b4dcfc1fc348d8234d0c5db158962043720828be2eb9085 clauseid=d3cdb02e-6982-47a1-9ba3-350d46212b63>
Acceptance of Delivery. <variable id=\\"shipper\\" value=\\"%22Party%20A%22\\"/> will be deemed to have completed its delivery obligations if in <variable id=\\"receiver\\" value=\\"%22Party%20B%22\\"/>'s opinion, the <variable id=\\"deliverable\\" value=\\"%22Widgets%22\\"/> satisfies the Acceptance Criteria, and <variable id=\\"receiver\\" value=\\"%22Party%20B%22\\"/> notifies <variable id=\\"shipper\\" value=\\"%22Party%20A%22\\"/> in writing that it is accepting the <variable id=\\"deliverable\\" value=\\"%22Widgets%22\\"/>.

Inspection and Notice. <variable id=\\"receiver\\" value=\\"%22Party%20B%22\\"/> will have <variable id=\\"businessDays\\" value=\\"10\\"/> Business Days' to inspect and evaluate the <variable id=\\"deliverable\\" value=\\"%22Widgets%22\\"/> on the delivery date before notifying <variable id=\\"shipper\\" value=\\"%22Party%20A%22\\"/> that it is either accepting or rejecting the <variable id=\\"deliverable\\" value=\\"%22Widgets%22\\"/>.

Acceptance Criteria. The \\"Acceptance Criteria\\" are the specifications the <variable id=\\"deliverable\\" value=\\"%22Widgets%22\\"/> must meet for the <variable id=\\"shipper\\" value=\\"%22Party%20A%22\\"/> to comply with its requirements and obligations under this agreement, detailed in <variable id=\\"attachment\\" value=\\"%22Attachment%20X%22\\"/>, attached to this agreement.
\`\`\`
"
`;
